core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -98,6 +99,22 @@ private boolean checkTasks() {
 
       for (int i = 0; i < taskFutures.length; i += 1) {
         if (taskFutures[i] == null || taskFutures[i].isDone()) {
+          if (taskFutures[i] != null) {
+            // check for task failure and re-throw any exception
+            try {
+              taskFutures[i].get();
```

Comment:
So I see that there are multple places where ParallelIterator class is `synchronized` or otherwise blocking. However, from my reading of the rest of the class at least, it seems that we can submit two tasks per worker.

Is there any way we can avoid blocking on futures indefinitely in a loop, like for example possibly somehow joining the currently running futures and then throwing as sooon as one of them throws an exception. Perhaps there's something I'm missing that makes this less of a concern than it seems initially. If that's the case, could we possibly add a comment indicating why it's not a concern?

I'm just thinking out loud here, but to me, Future.get usually feels like a code smell.

Though I'm supportive if this will fix flaky tests. And if there's no concern, then that's great. If there is concern, my feeling is that maybe it's best to fix the flaky tests and then open an issue to be sure to come back to this to have `ParallelIterable` be more... well, parallel. But again, I might be missing something that makes it more parallel than it seems to me at first glance.

Reply:
Possibly CompletableFuture would make sense here?

Reply:
This class is used to parallelize operations that are consumed from a single thread, and primarily for scan planning where we need to scan lots of manifest files (the bottleneck) but split and combine them into tasks (quick). We want the reads to operate in parallel, but the consumer thread should not have additional concerns.

In addition, this class attempts to avoid running the back-end tasks too quickly and accumulating too many results in memory. If the consumer stops calling `next`, then we should also not submit more tasks. This avoids large memory costs in scan planning for limit queries in Presto.

To avoid submitting tasks that are unnecessary and to avoid using an additional monitor thread, this class manages task submission in the `Iterator` methods called by the consumer. Those methods are synchronized so that monitoring the task futures is a safe operation, which should not add much overhead because we expect a single consumer thread.

I'm not quite sure why you'd think that `Future.get` is a code smell. Isn't that the correct way to get the result of the future?

Reply:
It is definitely the appropriate way to get the result from a future, but I'm used to blocking on all of them or on a batch of them at one time (say 2 or so at one time) via some kind of join. However, this is absolutly just because I reviewed this when I was too tired. I should have definitely waited until I had fresh eyes to consider. I completely overlooked the part where you have `taskFutures[i].isDone()`. So that's totally my bad.

But now that you explain how this class is utilized, in particular that the consumer is single threaded, I better understand why this is written the way it is. 

I do very much appreciate you taking the time to explain to me how the class is utilized:
- The single threaded consumer
- The avoidance of large memory costs in scan planning for presto, particularly with limit queries.
- As well as the fact that the consumer might stop calling next.

I had mistakenly assumed that by this point in time, the iterator would definitely iterate over all manifests.

To further explain what I had thinking (since you graciously took the time to explain this classes utilization in detail), what I would have expected would be to loop over two of them or N of them, then await (via a join or some call to .get) on N at once. But looking at it again after getting more rest, I can see where I was mistaken. Especially since you have the check for `taskFutures[i].isDone()`. So I do apologize for leaving this review when I should have probably slept on it and then looked at it again with fresh eyes.

Again, thanks for taking the time to help me better understand this classes utilization and the unique concerns of various query engines. I don't have much presto experience, so this was definitely enlightening.

Thanks @rdblue!

Reply:
You can also mark this comment thread as outdated or otherwise close it how you want. I don't seem to have the ability to do that.